pkg/adaptor/controller.go

```diff
@@ -0,0 +1,365 @@
+			if err != nil {
+				return err
+			}
+			log.Infof("updated serviceprofile/%s as it's not equivalent to trafficsplit/%s", sp.Name, ts.Name)
+		}
+	}
+
+	return nil
+}
+
+func ignoreAnnotationPresent(sp *serviceprofile.ServiceProfile) bool {
+	_, ok := sp.Annotations[ignoreServiceProfileAnnotation]
+	return ok
+}
+
+func equal(spA *serviceprofile.ServiceProfile, spB *serviceprofile.ServiceProfile) bool {
+	if spA.Name != spB.Name {
+		return false
+	}
+
+	if spA.Namespace != spB.Namespace {
+		return false
+	}
+
+	if len(spA.Spec.DstOverrides) != len(spB.Spec.DstOverrides) {
+		return false
+	}
+
+	dstOverridesA := make(map[string]string)
+	for _, dstA := range spA.Spec.DstOverrides {
+		dstOverridesA[dstA.Authority] = dstA.Weight.String()
+	}
+
+	// Check if all the authorties from spB exist
+	// in dstOverridesA with the same weight
+	for _, dstB := range spB.Spec.DstOverrides {
+		weight, ok := dstOverridesA[dstB.Authority]
+		if !ok {
+			return false
+		}
+
+		if weight != dstB.Weight.String() {
+			return false
+		}
+	}
+
+	return true
+}
+
+// enqueueTS takes a Ts resource and converts it into a key
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than TS.
+func (c *SMIController) enqueueTS(obj interface{}) {
+	var key string
+	var err error
+	if key, err = trafficSplitKeyFunc(obj); err != nil {
+		log.Error(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+// trafficSplitKeyFunc takes a TS, and gives back a
+// namespace/name/service key string
+func trafficSplitKeyFunc(obj interface{}) (string, error) {
+	ts, ok := obj.(*trafficsplit.TrafficSplit)
+	if !ok {
+		return "", fmt.Errorf("couldn't convert the object in the queue to a trafficsplit")
+	}
+
+	return ts.Namespace + "/" + ts.Name + "/" + ts.Spec.Service, nil
+}
+
+func splitTrafficSplitKey(key string) (namespace, name, service string, err error) {
+	parts := strings.Split(key, "/")
+	if len(parts) != 3 {
+		return "", "", "", fmt.Errorf("unexpected key format: %q", key)
+	}
+
+	return parts[0], parts[1], parts[2], nil
+}
+
+// updateDstOverrides updates the dstOverrides of the given serviceprofile
+// to match that of the trafficsplit
+func updateDstOverrides(sp *serviceprofile.ServiceProfile, ts *trafficsplit.TrafficSplit, clusterDomain string) {
+	sp.Spec.DstOverrides = []*serviceprofile.WeightedDst{}
+	for _, backend := range ts.Spec.Backends {
+		weightedDst := &serviceprofile.WeightedDst{
+			Authority: fqdn(backend.Service, ts.Namespace, clusterDomain),
+			Weight:    *backend.Weight,
+		}
+		sp.Spec.DstOverrides = append(sp.Spec.DstOverrides, weightedDst)
+	}
+
+}
+
+// toServiceProfile converts the given TrafficSplit into the relevant ServiceProfile resource
+func (c *SMIController) toServiceProfile(ts *trafficsplit.TrafficSplit) *serviceprofile.ServiceProfile {
+	spResource := serviceprofile.ServiceProfile{
```

Comment:
I think you need to set Routes to an empty list (instead of nil) in order to avoid validation errors:

```
time="2021-06-09T20:45:25Z" level=error msg="error syncing 'booksapp/error-split/books': ServiceProfile.linkerd.io \"books.booksapp.svc.cluster.local\" is invalid: spec.routes: Invalid value: \"null\": spec.routes in body must be of type array: \"null\", requeuing"
```

Reply:
Yep, That is the case on stable, but that restriction is removed on the edge. Are you testing on the edge? 🤔 

I've tried to not touch other fields except for `dstOverrides` to keep the logic simple. Based on our discussions about supporting < `2.10.2`, I can work on fixing this.